src/main/java/org/truffleruby/language/exceptions/RescueNodeWrapper.java

```diff
@@ -11,77 +11,64 @@
 
-    @Override
-    public WrapperNode createWrapper(RescueNode delegateNode, ProbeNode probeNode) {
-        return new RescueNodeWrapper0(delegateNode, probeNode);
-    }
-
-    private static final class RescueNodeWrapper0 extends RescueNode implements WrapperNode {
-
-        @Child private RescueNode delegateNode;
-        @Child private ProbeNode probeNode;
+    @Child private RescueNode delegateNode;
+    @Child private ProbeNode probeNode;
 
-        private RescueNodeWrapper0(RescueNode delegateNode, ProbeNode probeNode) {
-            super(null);
-            this.delegateNode = delegateNode;
-            this.probeNode = probeNode;
-        }
+    RescueNodeWrapper(RescueNode delegateNode, ProbeNode probeNode) {
+        super(null);
+        this.delegateNode = delegateNode;
+        this.probeNode = probeNode;
+    }
 
-        @Override
-        public Node getDelegateNode() {
-            return delegateNode;
-        }
+    @Override
+    public RescueNode getDelegateNode() {
+        return delegateNode;
+    }
 
-        @Override
-        public ProbeNode getProbeNode() {
-            return probeNode;
-        }
+    @Override
+    public ProbeNode getProbeNode() {
+        return probeNode;
+    }
 
-        @Override
-        public NodeCost getCost() {
-            return NodeCost.NONE;
-        }
+    @Override
+    public NodeCost getCost() {
+        return NodeCost.NONE;
+    }
 
-        @SuppressWarnings("deprecation")
-        @Override
-        public Object execute(VirtualFrame frame) {
+    @Override
+    public Object execute(VirtualFrame frame) {
+        Object returnValue;
+        for (;;) {
+            boolean wasOnReturnExecuted = false;
             try {
                 probeNode.onEnter(frame);
-                Object returnValue = delegateNode.execute(frame);
+                returnValue = delegateNode.execute(frame);
+                wasOnReturnExecuted = true;
                 probeNode.onReturnValue(frame, returnValue);
-                return returnValue;
-            } catch (Throwable t) {
-                probeNode.onReturnExceptional(frame, t);
-                throw t;
-            }
-        }
-
-        @SuppressWarnings("deprecation")
-        @Override
-        public void doExecuteVoid(VirtualFrame frame) {
-            try {
-                probeNode.onEnter(frame);
-                delegateNode.doExecuteVoid(frame);
-                probeNode.onReturnValue(frame, null);
+                break;
             } catch (Throwable t) {
-                probeNode.onReturnExceptional(frame, t);
+                Object result = probeNode.onReturnExceptionalOrUnwind(frame, t, wasOnReturnExecuted);
+                if (result == ProbeNode.UNWIND_ACTION_REENTER) {
+                    continue;
+                } else if (result != null) {
+                    returnValue = result;
+                    break;
+                }
```

Comment:
Is this a documented pattern or that same as the generated wrapper?
Looks like `return` could be used in the loop directly instead of `returnValue` (then it can a local variable only in the `try` part).

Reply:
It's the same as the generated code. It would be nice if we didn't have to manually generate the wrapper here, but we need to for `canHandle()`.

Reply:
Can you add a comment about why this wrapper is needed at the top of the class?